File: 6-Data_structures_arrays/merge_sorted_arrays.rb
# frozen_string_literal: true

def merge_sorted_arrays(array1, array2)
  merged_arr = []
  i = 0
  j = 0

  array2 if array1.length.zero?
  array1 if array2.length.zero?

  while array1[i] || array2[j]
    if !array1[i].nil? && (array2[j].nil? || array1[i] <= array2[j])
      merged_arr << array1[i]
      i += 1
    else
      merged_arr << array2[j]
      j += 1
    end
  end

  merged_arr

  # until array1.empty? || array2.empty?
  #   merged_arr << if array1.first <= array2.first
  #                   array1.shift
  #                 else
  #                   array2.shift
  #                 end
  # end

  # merged_arr + array1 + array2

  # (array1 + array2).sort
end

p merge_sorted_arrays([0, 3, 4, 31], [3, 4, 6, 30])

File: 6-Data_structures_arrays/merge_sorted_arrays.js
const mergeSortedArrays = (array1, array2) => {
  const mergedArray = [];
  let array1Item = array1[0];
  let arrat2Item = array2[0];
  let i = 1;
  let j = 1;

  if (array1.length === 0) return array2;
  if (array2.length === 0) return array1;

  while (array1Item || arrat2Item) {
    if (arrat2Item === undefined || array1Item < arrat2Item) {
      mergedArray.push(array1Item);
      array1Item = array1[i];
      i++;
    } else {
      mergedArray.push(arrat2Item);
      arrat2Item = array2[j];
      j++;
    }
  }

  console.log(mergedArray);

  // return [...array1, ...array2].sort((a, b) => a - b);
};

console.log(mergeSortedArrays([0, 3, 4, 31], [3, 4, 6, 30]));
